File: weird_captcha_gym/shared_runtime/server/incubator_graders/letter_rapids.py
from __future__ import annotations

from typing import Any


MECHANIC_ID = "letter_rapids"


def _failure(message: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "score": 0, "feedback": message}


def _integer(value: Any) -> int | None:
    if isinstance(value, bool):
        return None
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def _identity_error(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> str | None:
    if {
        str(payload.get("mechanic_id") or ""),
        str(ground_truth.get("mechanic_id") or ""),
        str(public_state.get("mechanic_id") or ""),
    } != {MECHANIC_ID}:
        return "mechanic identity mismatch"
    challenges = {
        str(payload.get("challenge_id") or ""),
        str(ground_truth.get("challenge_id") or ""),
        str(public_state.get("challenge_id") or ""),
    }
    if len(challenges) != 1 or "" in challenges:
        return "challenge identity mismatch"
    tasks = {
        str(payload.get("task_id") or ""),
        str(ground_truth.get("task_id") or ""),
        str(public_state.get("task_id") or ""),
    }
    if len(tasks) != 1 or "" in tasks:
        return "task identity mismatch"
    return None


def _display_row(row: list[dict[str, Any]], floor: int) -> list[tuple[dict[str, Any], int, int]]:
    free = 10_000 - floor * len(row)
    widths = [
        floor + free * (int(band["end_milli"]) - int(band["start_milli"])) // 10_000
        for band in row
    ]
    remainder = 10_000 - sum(widths)
    for index in range(remainder):
        widths[index] += 1
    cursor = 0
    result = []
    for band, width in zip(row, widths):
        result.append((band, cursor, cursor + width))
        cursor += width
    return result


def _symbol_at(rows: dict[str, Any], output: str, y_milli: int, display_floor: int) -> str | None:
    context = output[-1] if output else "^"
    row = rows.get(context)
    if not isinstance(row, list):
        return None
    for band, start, end in _display_row(row, display_floor):
        if not isinstance(band, dict):
            return None
        if start <= y_milli < end:
            return str(band.get("symbol") or " ")
    return None


def _flow_delta(x_milli: int, current_milli: int, simulation: dict[str, Any]) -> tuple[int, int]:
    neutral = int(simulation["neutral_x_milli"])
    dead = int(simulation["dead_zone_half_width_milli"])
    speed = int(simulation["maximum_speed_units_per_second"])
    tick_ms = int(simulation["tick_ms"])
    forward_edge = neutral + dead
    reverse_edge = neutral - dead
    if x_milli > forward_edge:
        distance = x_milli - forward_edge
        extent = 10_000 - forward_edge
        delta = speed * tick_ms * distance * current_milli // (1_000 * extent * 1_000)
        return 1, delta
    if x_milli < reverse_edge:
        distance = reverse_edge - x_milli
        extent = reverse_edge
        delta = speed * tick_ms * distance * current_milli // (1_000 * extent * 1_000)
        return -1, delta
    return 0, 0


def _contract_error(ground_truth: dict[str, Any], public_state: dict[str, Any]) -> str | None:
    keys = ("alphabet", "target", "probability_rows", "current_pattern_milli", "simulation")
    if any(public_state.get(key) != ground_truth.get(key) for key in keys):
        return "public canyon commitment disagrees with hidden state"
    alphabet = ground_truth.get("alphabet")
    target = ground_truth.get("target")
    rows = ground_truth.get("probability_rows")
    pattern = ground_truth.get("current_pattern_milli")
    simulation = ground_truth.get("simulation")
    if not isinstance(alphabet, str) or not alphabet or not isinstance(target, str) or not target:
        return "hidden letter contract is malformed"
    if not isinstance(rows, dict) or set(rows) != set("^" + alphabet):
        return "hidden probability model is malformed"
    expected_symbols = list(alphabet)
    for row in rows.values():
        if not isinstance(row, list) or [str(item.get("symbol") or " ") for item in row if isinstance(item, dict)] != expected_symbols:
            return "hidden probability row has the wrong alphabet"
        cursor = 0
        for band in row:
            if _integer(band.get("start_milli")) != cursor:
                return "hidden probability row has a gap"
            cursor = _integer(band.get("end_milli")) or -1
        if cursor != 10_000:
            return "hidden probability row does not fill the canyon"
    if not isinstance(pattern, list) or len(pattern) < 100 or any(_integer(value) is None for value in pattern):
        return "hidden current trace is malformed"
    required = {
        "tick_ms", "commit_units", "neutral_x_milli", "dead_zone_half_width_milli",
        "maximum_speed_units_per_second", "travel_budget_units", "maximum_rewound_characters",
        "display_band_floor_milli",
    }
    if not isinstance(simulation, dict) or not required.issubset(simulation):
        return "hidden simulation contract is malformed"
    display_floor = _integer(simulation.get("display_band_floor_milli"))
    if display_floor is None or not 200 <= display_floor <= 350 or display_floor * len(alphabet) >= 10_000:
        return "hidden display geometry is malformed"
    return None


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    identity_error = _identity_error(payload, ground_truth, public_state)
    if identity_error:
        return _failure(identity_error)
    contract_error = _contract_error(ground_truth, public_state)
    if contract_error:
        return _failure(contract_error)

    condition = ground_truth.get("control_condition")
    interaction = "full"
    if condition is not None:
        if public_state.get("control_condition") != condition:
            return _failure("controlled canyon condition mismatch")
        interaction = str(condition.get("interaction") or "")
    if interaction not in {"full", "simplified"} or payload.get("interaction_mode") != interaction:
        return _failure("wrong canyon interaction mode")
    expected_source = {"full": "canyon_pointer", "simplified": "axis_proxy"}[interaction]

    events = payload.get("events")
    if not isinstance(events, list) or not events or len(events) > 600:
        return _failure("pointer transcript is missing or oversized")
    parsed_events: list[tuple[int, int, int]] = []
    previous_tick = -1
    for index, event in enumerate(events, start=1):
        if not isinstance(event, dict) or _integer(event.get("seq")) != index:
            return _failure("pointer sequence is not contiguous")
        if event.get("type") != "pointer":
            return _failure("unknown canyon transcript event")
        tick = _integer(event.get("tick"))
        x_milli = _integer(event.get("x_milli"))
        y_milli = _integer(event.get("y_milli"))
        if tick is None or tick < previous_tick or x_milli is None or y_milli is None:
            return _failure("pointer event coordinates are malformed")
        if not 0 <= x_milli <= 10_000 or not 0 <= y_milli < 10_000:
            return _failure("pointer left the physical canyon range")
        if event.get("input_source") != expected_source:
            return _failure("pointer used the wrong interaction input")
        previous_tick = tick
        parsed_events.append((tick, x_milli, y_milli))

    terminal_tick = _integer(payload.get("terminal_tick"))
    pattern = ground_truth["current_pattern_milli"]
    if terminal_tick is None or not 1 <= terminal_tick <= len(pattern):
        return _failure("terminal simulation tick is invalid")
    if parsed_events[-1][0] >= terminal_tick:
        return _failure("pointer transcript continues beyond the terminal tick")

    simulation = ground_truth["simulation"]
    display_floor = int(simulation["display_band_floor_milli"])
    rows = ground_truth["probability_rows"]
    target = str(ground_truth["target"])
    x_milli = int(simulation["neutral_x_milli"])
    y_milli = 5_000
    output = ""
    selected_symbol: str | None = None
    progress = 0
    rewind_progress = 0
    travel_used = 0
    committed = 0
    rewound = 0
    terminal_reason: str | None = None
    event_index = 0

    for tick in range(terminal_tick):
        while event_index < len(parsed_events) and parsed_events[event_index][0] == tick:
            _, x_milli, y_milli = parsed_events[event_index]
            event_index += 1
        direction, delta = _flow_delta(x_milli, int(pattern[tick]), simulation)
        if direction > 0 and delta:
            symbol = _symbol_at(rows, output, y_milli, display_floor)
            if symbol is None:
                return _failure("probability hit test could not resolve a channel")
            if symbol != selected_symbol:
                selected_symbol = symbol
                progress = 0
            rewind_progress = 0
            progress += delta
            travel_used += delta
            if progress >= int(simulation["commit_units"]):
                output += symbol
                committed += 1
                progress = 0
                selected_symbol = None
                if output == target:
                    terminal_reason = "target"
        elif direction < 0 and delta:
            travel_used += delta
            remaining = delta
            cancelled = min(progress, remaining)
            progress -= cancelled
            remaining -= cancelled
            if remaining:
                rewind_progress += remaining
            while rewind_progress >= int(simulation["commit_units"]) and output:
                rewind_progress -= int(simulation["commit_units"])
                output = output[:-1]
                rewound += 1
                selected_symbol = None
            if rewound > int(simulation["maximum_rewound_characters"]):
                terminal_reason = "rewind_budget"
        if terminal_reason is None and travel_used >= int(simulation["travel_budget_units"]):
            terminal_reason = "travel_budget"
        if terminal_reason is not None and tick != terminal_tick - 1:
            return _failure("submission continued after the replay reached a terminal state")

    if terminal_reason is None:
        return _failure("submission stopped before a terminal state")
    if event_index != len(parsed_events):
        return _failure("unreplayed pointer events remain")

    claims = {
        "output": output,
        "terminal_reason": terminal_reason,
        "travel_used_units": travel_used,
        "committed_characters": committed,
        "rewound_characters": rewound,
        "progress_units": progress,
    }
    for key, expected in claims.items():
        actual = payload.get(key)
        if isinstance(expected, int):
            actual = _integer(actual)
        if actual != expected:
            return _failure(f"submitted {key.replace('_', ' ')} disagrees with replay")
    passed = payload.get("completed") is True and terminal_reason == "target" and output == target
    if payload.get("completed") is not (terminal_reason == "target"):
        return _failure("submitted completion flag disagrees with replay")
    return {
        "graded": True,
        "passed": passed,
        "score": 100 if passed else 0,
        "feedback": (
            f"replayed {terminal_tick} current ticks; output {output!r}/{target!r}; "
            f"commits {committed}; rewinds {rewound}; travel {travel_used}/{int(simulation['travel_budget_units'])}; "
            f"terminal {terminal_reason}"
        ),
    }


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    del public_state
    return {"target": ground_truth.get("target")}

File: weird_captcha_gym/shared_runtime/app/mechanics/letter_rapids.js
(() => {
  "use strict";

  let model = null;
  let cleanupActive = null;
  const COLORS = ["#21d6b1", "#f2bd55", "#ef6d6a", "#73a8ff", "#c7e66b", "#d98bff", "#55d6ee"];
  const esc = (value) => String(value == null ? "" : value)
    .replaceAll("&", "&amp;").replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;").replaceAll('"', "&quot;");

  function interactionMode() {
    return model?.state?.control_condition?.interaction || "full";
  }

  function isControlled() {
    return Boolean(model?.state?.control_condition);
  }

  function currentRow(output = model.output) {
    const context = output.length ? output.at(-1) : "^";
    return model.state.probability_rows[context];
  }

  function displayRow(output = model.output) {
    const row = currentRow(output);
    const floor = Number(model.state.simulation.display_band_floor_milli);
    const free = 10000 - floor * row.length;
    const widths = row.map((band) => floor + Math.floor(
      free * (Number(band.end_milli) - Number(band.start_milli)) / 10000
    ));
    let remainder = 10000 - widths.reduce((total, width) => total + width, 0);
    for (let index = 0; remainder > 0; index += 1, remainder -= 1) widths[index] += 1;
    let cursor = 0;
    return row.map((band, index) => {
      const visible = {...band, display_start_milli: cursor, display_end_milli: cursor + widths[index]};
      cursor += widths[index];
      return visible;
    });
  }

  function symbolAt(yMilli, output = model.output) {
    const band = displayRow(output).find((item) => Number(item.display_start_milli) <= yMilli && yMilli < Number(item.display_end_milli));
    return band ? String(band.symbol) : null;
  }

  function flowDelta(xMilli, currentMilli) {
    const sim = model.state.simulation;
    const neutral = Number(sim.neutral_x_milli);
    const dead = Number(sim.dead_zone_half_width_milli);
    const speed = Number(sim.maximum_speed_units_per_second);
    const tickMs = Number(sim.tick_ms);
    const forwardEdge = neutral + dead;
    const reverseEdge = neutral - dead;
    if (xMilli > forwardEdge) {
      const distance = xMilli - forwardEdge;
      const extent = 10000 - forwardEdge;
      return {direction: 1, delta: Math.floor(speed * tickMs * distance * currentMilli / (1000 * extent * 1000))};
    }
    if (xMilli < reverseEdge) {
      const distance = reverseEdge - xMilli;
      const extent = reverseEdge;
      return {direction: -1, delta: Math.floor(speed * tickMs * distance * currentMilli / (1000 * extent * 1000))};
    }
    return {direction: 0, delta: 0};
  }

  function recordPointer(xMilli, yMilli, inputSource) {
    if (!model || model.submitting || model.terminalReason) return;
    const x = Math.max(0, Math.min(10000, Math.round(Number(xMilli))));
    const y = Math.max(0, Math.min(9999, Math.round(Number(yMilli))));
    // A release from the control that submitted the previous failed route can
    // land on the freshly rendered equivalent control.  Ignore coordinate-
    // identical input unconditionally so that this trailing release cannot
    // dismiss the visible failure state or add an event to the new route.
    if (x === model.pointer.x && y === model.pointer.y) return;
    model.pointer = {x, y};
    const event = {seq: model.events.length + 1, tick: model.tick, type: "pointer", x_milli: x, y_milli: y};
    event.input_source = inputSource;
    model.events.push(event);
    clearFreshFailure();
    updateHud();
  }

  function clearFreshFailure() {
    const root = document.querySelector(".letter-rapids");
    const wasFresh = root?.hasAttribute("data-fresh-failure");
    root?.removeAttribute("data-fresh-failure");
    root?.querySelector(".rapids-fresh-stamp")?.remove();
    if (wasFresh) model?.helpers.setReadout("READY", "idle");
  }

  function finish(reason) {
    if (model.terminalReason) return;
    model.terminalReason = reason;
    model.helpers.setReadout("CHECKING", "pending");
    updateHud();
    queueMicrotask(submitTerminal);
  }

  function stepSimulation() {
    if (!model || model.terminalReason) return;
    const sim = model.state.simulation;
    const pattern = model.state.current_pattern_milli;
    if (model.tick >= pattern.length) {
      finish("travel_budget");
      return;
    }
    const {direction, delta} = flowDelta(model.pointer.x, Number(pattern[model.tick]));
    if (direction > 0 && delta) {
      const symbol = symbolAt(model.pointer.y);
      if (symbol !== model.selectedSymbol) {
        model.selectedSymbol = symbol;
        model.progress = 0;
      }
      model.rewindProgress = 0;
      model.progress += delta;
      model.travelUsed += delta;
      if (model.progress >= Number(sim.commit_units)) {
        model.output += symbol;
        model.committed += 1;
        model.progress = 0;
        model.selectedSymbol = null;
        model.commitFlash = 8;
        if (model.output === model.state.target) model.pendingReason = "target";
      }
    } else if (direction < 0 && delta) {
      model.travelUsed += delta;
      let remaining = delta;
      const cancelled = Math.min(model.progress, remaining);
      model.progress -= cancelled;
      remaining -= cancelled;
      if (remaining) model.rewindProgress += remaining;
      while (model.rewindProgress >= Number(sim.commit_units) && model.output.length) {
        model.rewindProgress -= Number(sim.commit_units);
        model.output = model.output.slice(0, -1);
        model.rewound += 1;
        model.selectedSymbol = null;
        model.commitFlash = -8;
      }
      if (model.rewound > Number(sim.maximum_rewound_characters)) model.pendingReason = "rewind_budget";
    }
    if (!model.pendingReason && model.travelUsed >= Number(sim.travel_budget_units)) model.pendingReason = "travel_budget";
    model.tick += 1;
    if (model.commitFlash > 0) model.commitFlash -= 1;
    else if (model.commitFlash < 0) model.commitFlash += 1;
    updateHud();
    if (model.pendingReason) finish(model.pendingReason);
  }

  async function submitTerminal() {
    if (!model || model.submitting || !model.terminalReason) return;
    model.submitting = true;
    const payload = {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      events: model.events,
      completed: model.terminalReason === "target",
      output: model.output,
      terminal_reason: model.terminalReason,
      terminal_tick: model.tick,
      travel_used_units: model.travelUsed,
      committed_characters: model.committed,
      rewound_characters: model.rewound,
      progress_units: model.progress,
    };
    payload.interaction_mode = interactionMode();
    try {
      const response = await fetch("/result", {method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify(payload)});
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.submitting = false;
        document.querySelector(".letter-rapids")?.setAttribute("data-terminal", "pass");
        document.querySelector(".letter-rapids")?.insertAdjacentHTML("beforeend", '<div class="rapids-terminal"><strong>PASS</strong></div>');
        model.helpers.setReadout("PASS", "passed");
      } else if (outcome.passed === false && outcome.state) {
        await render(outcome.state, model.helpers, {freshFailure: true});
        model.helpers.setReadout("FAIL", "error");
      } else {
        model.submitting = false;
        model.helpers.setReadout("VERIFICATION LINK UNAVAILABLE", "error");
      }
    } catch (_error) {
      model.submitting = false;
      model.helpers.setReadout("VERIFICATION LINK LOST", "error");
    }
  }

  function printable(symbol) {
    return symbol === " " ? "SP" : symbol.toUpperCase();
  }

  function updateHud() {
    if (!model) return;
    const symbol = symbolAt(model.pointer.y) || "?";
    const flow = flowDelta(model.pointer.x, Number(model.state.current_pattern_milli[Math.min(model.tick, model.state.current_pattern_milli.length - 1)]));
    const mode = flow.direction > 0 ? "ADVANCE" : flow.direction < 0 ? "REWIND" : "BRAKE";
    const output = document.querySelector(".rapids-output-value");
    if (output) {
      output.textContent = model.output || "—";
      output.dataset.output = model.output;
    }
    const selected = document.querySelector(".rapids-selected-letter");
    if (selected) selected.textContent = printable(symbol);
    const helm = document.querySelector(".rapids-helm-mode");
    if (helm) { helm.textContent = mode; helm.dataset.mode = mode.toLowerCase(); }
    const distance = document.querySelector(".rapids-distance i");
    if (distance) distance.style.width = `${Math.min(100, model.travelUsed / Number(model.state.simulation.travel_budget_units) * 100)}%`;
    const progress = document.querySelector(".rapids-gate-progress i");
    if (progress) progress.style.width = `${Math.min(100, model.progress / Number(model.state.simulation.commit_units) * 100)}%`;
    const current = document.querySelector(".rapids-current-number");
    if (current) current.textContent = `${(Number(model.state.current_pattern_milli[Math.min(model.tick, model.state.current_pattern_milli.length - 1)]) / 1000).toFixed(2)}×`;
    const rewind = document.querySelector(".rapids-rewind-count");
    if (rewind) rewind.textContent = `${model.rewound} / ${model.state.simulation.maximum_rewound_characters}`;
    const aim = document.getElementById("rapids-aim");
    if (aim && Number(aim.value) !== model.pointer.y) aim.value = String(model.pointer.y);
    const proxyFlow = document.getElementById("rapids-flow");
    if (proxyFlow && Number(proxyFlow.value) !== model.pointer.x) proxyFlow.value = String(model.pointer.x);
  }

  function roundedRect(ctx, x, y, width, height, radius) {
    const r = Math.min(radius, Math.abs(width) / 2, Math.abs(height) / 2);
    ctx.beginPath();
    ctx.roundRect(x, y, width, height, r);
  }

  function drawCanvas() {
    if (!model?.canvas) return;
    const canvas = model.canvas;
    const ctx = model.context;
    const width = canvas.width;
    const height = canvas.height;
    const sim = model.state.simulation;
    const neutralX = Number(sim.neutral_x_milli) / 10000 * width;
    const reverseEdge = (Number(sim.neutral_x_milli) - Number(sim.dead_zone_half_width_milli)) / 10000 * width;
    const forwardEdge = (Number(sim.neutral_x_milli) + Number(sim.dead_zone_half_width_milli)) / 10000 * width;
    const pointerX = model.pointer.x / 10000 * width;
    const pointerY = model.pointer.y / 10000 * height;
    const current = Number(model.state.current_pattern_milli[Math.min(model.tick, model.state.current_pattern_milli.length - 1)]);

    const background = ctx.createLinearGradient(0, 0, width, height);
    background.addColorStop(0, "#08191b");
    background.addColorStop(.52, "#0b2527");
    background.addColorStop(1, "#061416");
    ctx.fillStyle = background;
    ctx.fillRect(0, 0, width, height);

    const row = displayRow();
    const active = symbolAt(model.pointer.y);
    row.forEach((band, index) => {
      const y0 = Number(band.display_start_milli) / 10000 * height;
      const y1 = Number(band.display_end_milli) / 10000 * height;
      const selected = band.symbol === active;
      ctx.fillStyle = COLORS[index % COLORS.length] + (selected ? "45" : "1b");
      ctx.fillRect(0, y0, width, Math.max(1, y1 - y0));
      ctx.strokeStyle = selected ? COLORS[index % COLORS.length] + "d8" : "rgba(209,247,237,.17)";
      ctx.lineWidth = selected ? 2 : 1;
      ctx.beginPath(); ctx.moveTo(0, y0 + .5); ctx.lineTo(width, y0 + .5); ctx.stroke();
      const bandHeight = y1 - y0;
      if (bandHeight >= 7) {
        ctx.fillStyle = selected ? "#f7fff8" : "rgba(224,249,242,.72)";
        ctx.font = `${bandHeight < 13 ? 8 : bandHeight < 21 ? 10 : 13}px ui-monospace, SFMono-Regular, Menlo, monospace`;
        ctx.textBaseline = "middle";
        ctx.fillText(printable(band.symbol), forwardEdge + 10, (y0 + y1) / 2);
      }
    });

    const activeBand = row.find((band) => band.symbol === active);
    if (activeBand) {
      const y0 = Number(activeBand.display_start_milli) / 10000 * height;
      const y1 = Number(activeBand.display_end_milli) / 10000 * height;
      const front = forwardEdge + (1 - model.progress / Number(sim.commit_units)) * (width - forwardEdge);
      const glow = ctx.createLinearGradient(front, 0, width, 0);
      glow.addColorStop(0, "rgba(255,255,255,.06)");
      glow.addColorStop(1, "rgba(255,255,255,.20)");
      ctx.fillStyle = glow;
      ctx.fillRect(front, y0, width - front, y1 - y0);
      ctx.strokeStyle = "rgba(255,255,255,.92)";
      ctx.lineWidth = 3;
      ctx.beginPath(); ctx.moveTo(front, y0); ctx.lineTo(front, y1); ctx.stroke();
      if (y1 - y0 > 30) {
        const previewRow = model.state.probability_rows[active];
        previewRow.forEach((child, index) => {
          const cy0 = y0 + Number(child.start_milli) / 10000 * (y1 - y0);
          const cy1 = y0 + Number(child.end_milli) / 10000 * (y1 - y0);
          ctx.fillStyle = COLORS[(index + 2) % COLORS.length] + "25";
          ctx.fillRect(width * .76, cy0, width * .24, Math.max(1, cy1 - cy0));
          ctx.strokeStyle = "rgba(255,255,255,.13)";
          ctx.strokeRect(width * .76, cy0, width * .24, Math.max(1, cy1 - cy0));
          if (cy1 - cy0 >= 9) {
            ctx.fillStyle = "rgba(255,255,255,.68)";
            ctx.font = "8px ui-monospace, SFMono-Regular, Menlo, monospace";
            ctx.fillText(printable(child.symbol), width * .77, (cy0 + cy1) / 2);
          }
        });
      }
    }

    ctx.fillStyle = "rgba(238,88,85,.11)";
    ctx.fillRect(0, 0, reverseEdge, height);
    ctx.fillStyle = "rgba(250,222,150,.10)";
    ctx.fillRect(reverseEdge, 0, forwardEdge - reverseEdge, height);
    ctx.save();
    ctx.beginPath(); ctx.rect(reverseEdge, 0, forwardEdge - reverseEdge, height); ctx.clip();
    ctx.strokeStyle = "rgba(255,225,166,.16)"; ctx.lineWidth = 8;
    for (let x = reverseEdge - height; x < forwardEdge + height; x += 20) {
      ctx.beginPath(); ctx.moveTo(x, 0); ctx.lineTo(x + height, height); ctx.stroke();
    }
    ctx.restore();
    ctx.strokeStyle = "rgba(255,225,166,.85)"; ctx.lineWidth = 2;
    ctx.beginPath(); ctx.moveTo(reverseEdge, 0); ctx.lineTo(reverseEdge, height); ctx.stroke();
    ctx.beginPath(); ctx.moveTo(forwardEdge, 0); ctx.lineTo(forwardEdge, height); ctx.stroke();
    ctx.strokeStyle = "rgba(255,255,255,.28)"; ctx.lineWidth = 1;
    ctx.beginPath(); ctx.moveTo(neutralX, 0); ctx.lineTo(neutralX, height); ctx.stroke();

    const particleSpeed = .8 + current / 850;
    ctx.fillStyle = "rgba(194,255,241,.30)";
    for (let index = 0; index < 34; index += 1) {
      const phase = (model.tick * particleSpeed * 5 + index * 79) % (width + 90);
      const x = width + 25 - phase;
      const y = (index * 137 + 31) % height;
      ctx.fillRect(x, y, 18 + (index % 4) * 8, 1);
    }

    ctx.fillStyle = "rgba(5,13,14,.82)";
    roundedRect(ctx, 10, 11, 118, 25, 4); ctx.fill();
    ctx.font = "700 10px ui-monospace, SFMono-Regular, Menlo, monospace";
    ctx.fillStyle = "#ef7a75"; ctx.fillText("← REWIND", 20, 24);
    ctx.fillStyle = "#f2cf7c"; ctx.fillText("BRAKE", reverseEdge + 10, 24);
    ctx.fillStyle = "#72e5ca"; ctx.fillText("ADVANCE →", forwardEdge + 10, 24);

    ctx.save();
    ctx.translate(pointerX, pointerY);
    ctx.strokeStyle = "#ffffff";
    ctx.fillStyle = "rgba(7,18,19,.72)";
    ctx.lineWidth = 2;
    ctx.shadowColor = "rgba(117,239,216,.8)";
    ctx.shadowBlur = 9;
    ctx.beginPath(); ctx.arc(0, 0, 10, 0, Math.PI * 2); ctx.fill(); ctx.stroke();
    ctx.beginPath(); ctx.moveTo(-17, 0); ctx.lineTo(-6, 0); ctx.moveTo(6, 0); ctx.lineTo(17, 0); ctx.moveTo(0, -17); ctx.lineTo(0, -6); ctx.moveTo(0, 6); ctx.lineTo(0, 17); ctx.stroke();
    ctx.restore();

    model.raf = requestAnimationFrame(drawCanvas);
  }

  function animationLoop() {
    if (!model) return;
    const now = model.helpers.interactionNow();
    if (model.lastClock == null) model.lastClock = now;
    const elapsed = Math.max(0, now - model.lastClock);
    model.lastClock = now;
    model.accumulator += elapsed;
    while (model.accumulator >= Number(model.state.simulation.tick_ms) && !model.terminalReason) {
      model.accumulator -= Number(model.state.simulation.tick_ms);
      stepSimulation();
    }
    model.loopRaf = requestAnimationFrame(animationLoop);
  }

  function bindFullPointer() {
    const canvas = model.canvas;
    const update = (event) => {
      const rect = canvas.getBoundingClientRect();
      recordPointer((event.clientX - rect.left) / rect.width * 10000, (event.clientY - rect.top) / rect.height * 10000, "canyon_pointer");
    };
    canvas.addEventListener("pointermove", update);
  }

  function bindSimplifiedPointer() {
    const aim = document.getElementById("rapids-aim");
    const flow = document.getElementById("rapids-flow");
    aim.addEventListener("input", () => recordPointer(model.pointer.x, Number(aim.value), "axis_proxy"));
    flow.addEventListener("input", () => recordPointer(Number(flow.value), model.pointer.y, "axis_proxy"));
  }

  async function render(state, helpers, options = {}) {
    if (cleanupActive) cleanupActive();
    document.body.dataset.mechanic = "letter-rapids";
    document.body.dataset.cheatMode = helpers.isCheatMode() ? "true" : "false";
    const interaction = state.control_condition?.interaction || "full";
    const sim = state.simulation;
    model = {
      state, helpers, events: [], tick: 0, accumulator: 0, lastClock: null,
      pointer: {x: Number(sim.neutral_x_milli), y: 5000},
      output: "", selectedSymbol: null, progress: 0, rewindProgress: 0,
      travelUsed: 0, committed: 0, rewound: 0, pendingReason: null,
      terminalReason: null, submitting: false, commitFlash: 0,
      canvas: null, context: null, raf: 0, loopRaf: 0,
    };
    window.letterRapidsModel = model;
    const controlPanel = interaction === "simplified" ? `
      <section class="rapids-proxy" aria-label="simplified pointer controls">
        <label for="rapids-flow"><span>FLOW</span></label>
        <input id="rapids-flow" type="range" min="0" max="10000" step="1" value="${Number(sim.neutral_x_milli)}" aria-label="flow position">
        <div class="rapids-flow-poles" aria-hidden="true"><span>←</span><i></i><span>→</span></div>
      </section>` : "";
    const aimRail = interaction === "simplified" ? `
      <div class="rapids-aim-rail">
        <label for="rapids-aim">LETTER</label>
        <input id="rapids-aim" type="range" min="0" max="9999" step="1" value="5000" aria-label="letter position">
      </div>` : "";
    helpers.app.innerHTML = `
      <section class="letter-rapids" data-challenge-id="${esc(state.challenge_id)}" data-interaction="${esc(interaction)}" ${options.freshFailure ? 'data-fresh-failure="true"' : ""}>
        ${options.freshFailure ? '<div class="rapids-fresh-stamp"><b>FAIL</b></div>' : ""}
        <header class="rapids-head">
          <div><span>FLOOD CONTROL / LETTER LOCK 04</span><h1>LETTER RAPIDS</h1><p>${esc(state.prompt)}</p></div>
          <section class="rapids-target"><small>TARGET PHRASE</small><div>${[...state.target].map((symbol) => `<b class="rapids-target-char">${esc(printable(symbol))}</b>`).join("")}</div></section>
          <div class="rapids-live"><i></i><span>LIVE CURRENT</span></div>
        </header>
        <main class="rapids-main">
          <section class="rapids-canyon-shell"><canvas id="rapids-canvas" width="920" height="450" aria-label="flowing probability-sized letter canyon"></canvas><div class="rapids-glass"></div>${aimRail}</section>
          <aside class="rapids-console">
            <section class="rapids-selection"><small>SELECTED CHANNEL</small><strong class="rapids-selected-letter">?</strong></section>
            <section class="rapids-output"><small>LOCK OUTPUT</small><b class="rapids-output-value" data-output="">—</b></section>
            <section class="rapids-instrument"><div><span>HELM</span><b class="rapids-helm-mode" data-mode="brake">BRAKE</b></div><div><span>CURRENT</span><b class="rapids-current-number">1.00×</b></div></section>
            <section class="rapids-meter"><header><span>TRAVEL LEDGER</span><b>LIMIT</b></header><em class="rapids-distance"><i></i></em></section>
            <section class="rapids-rewind"><span>CHARACTER REWINDS</span><b class="rapids-rewind-count">0 / ${Number(sim.maximum_rewound_characters)}</b></section>
            ${controlPanel}
          </aside>
        </main>
        <footer class="rapids-foot"><div class="readout" data-status="idle">READY</div></footer>
        ${helpers.cheatPanelTemplate()}
      </section>`;
    model.canvas = document.getElementById("rapids-canvas");
    model.context = model.canvas.getContext("2d");
    if (interaction === "simplified") bindSimplifiedPointer(); else bindFullPointer();
    helpers.installCheatPanel();
    updateHud();
    model.raf = requestAnimationFrame(drawCanvas);
    model.loopRaf = requestAnimationFrame(animationLoop);
    cleanupActive = () => {
      if (model?.raf) cancelAnimationFrame(model.raf);
      if (model?.loopRaf) cancelAnimationFrame(model.loopRaf);
    };
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.letter_rapids = {rootSelector: ".letter-rapids", render};
})();
